

```{r}
library(pacman) 
p_load(data.table, dplyr, ggplot2, stringr, foreach, scales, ggpubr, RColorBrewer, circlize,
  ComplexHeatmap, tximport, readr, DESeq2, DT, clusterProfiler, xlsx, rtracklayer, ggrepel)
options(scipen=999)
#source("./code/functions.R")
knitr::opts_chunk$set(fig.width=9, fig.height=5)
dir.create(outdatadir, recursive=TRUE)

pval <- 0.0005
FC <- 1.5
d <- readRDS(paste0("output/code/DE_analysis_edgeR/", "/significant.RDS"))
```

Differentially expressed genes were identified using DESeq2 packages (adjusted p-value threshold was `r pval`  and LFC threshold was `r FC`). In total `r nrow(d)` differentially expressed genes were identified. `r nrow(d[logFC>0, ])` genes showed increased expression and  `r nrow(d[logFC<0, ])` decreased. 


```{r}
pr <- prcomp(t(d))
p <- pr$x  %>% 
  as.data.table(., keep.rownames=TRUE) %>% 
  .[, .(rn, PC1, PC2)] %>% 
  .[, sample :=  str_split(rn, "_") %>% sapply(., `[`, 1)] %>%
  .[, replicate :=  str_split(rn, "_") %>% sapply(., `[`, 2)] %>% 
  ggplot(aes(PC1, PC2, colour=sample, shape=replicate)) +
    geom_point(size=3) +
    xlab(paste0("PC1, ", round(summary(pr)$importance[2, 1]*100, 2), "%")) +
    ylab(paste0("PC2, ", round(summary(pr)$importance[2, 2]*100, 2), "%")) +
    theme_bw() +
    scale_color_manual(values=c(brewer.pal(n=3,"Set1")[1], brewer.pal(n=3,"Set1")[3])) +
     guides(color = guide_legend(ncol = 1), 
            shape = guide_legend(ncol = 1) ) +
    theme(legend.position ="bottom", 
          legend.title = element_text(size = 6),  
          axis.text = element_text(size = 6),  
          axis.title = element_text(size = 6),  
          legend.text = element_text(size = 4)) 
p
pdf(paste0(outdatadir, "PCA.pdf"), width=2.5, height=2.5)
  p 
dev.off()  
```

```{r}
d <- readRDS(paste0("output/code/DE_analysis_edgeR/", "/results.RDS"))

## gff with annotation and sRNAS
gff <- rtracklayer::import("NC_009004_with_sRNA.gff")
sarasas <- fread("input/qPCR_ID.txt", header=FALSE)
vardai <- fread("input/qPCR_ID_long.txt", header=FALSE)

tikslines <- merge(sarasas, vardai, by="V1", all=TRUE) %>% 
  setnames(., c("LocusTag", "Gene", "Set", "oldLocus", "fullGene")) %>% 
  .[is.na(oldLocus), fullGene := Gene] %>% 
  .[is.na(oldLocus), oldLocus := Gene] %>%
  .[grepl("sLLM1042", Set)]  %>% 
  .[fullGene != "alr"]

set.seed(2000)
p <- d %>% 
  as.data.table() %>%
  .[, .(rn, logFC, FDR)] %>%
  .[, yval := (-1) * log10(FDR)] %>%
  .[, type := "Non. sig."] %>%
  .[abs(logFC) >= log2(FC), type := "Non. sig."] %>%
  .[FDR <= pval, type := "Non. sig."] %>%
  .[(abs(logFC) >= log2(FC)) & (FDR <= pval), type := "FC & p-adj."] %>%
  .[, type := factor(type, levels=c("Non. sig.", "FC & p-adj."))] %>% 
  merge(., tikslines[, .(LocusTag, fullGene)], by.x="rn", by.y="LocusTag", all.x=TRUE)  %>% 
  ggplot(aes(logFC, yval, colour=type, label=fullGene)) +
    geom_point(size=0.25) +
    theme_bw() +
    geom_text_repel(
                    box.padding = 0.05,  
                    min.segment.length = unit(0, 'lines'), 
                    colour = "black", 
                    size = 2,
                    segment.size = 0.25, 
                    nudge_y = -0.5,    
                    nudge_x = 0) +
    ylab("-log10(padj)") +
    xlab("log2FoldChange") + 
    scale_colour_manual(values=c(brewer.pal(n=3,"Dark2")[8], brewer.pal(n=8,"Dark2")[1])) +
    #scale_shape_manual(values=c(17, 16)) +
    guides(color = guide_legend(ncol = 2)) +
    #coord_cartesian(xlim=c(-5.5, 5.6)) +
    scale_shape_manual(values=c(17, 16)) +
    guides(color = guide_legend(nrow = 2), shape = guide_legend(nrow = 2)) +
    theme(legend.position ="none", 
      axis.text = element_text(size = 8, colour="black"),  
      axis.title = element_text(size = 8, colour="black"),  
      plot.margin = margin(t = 1, l = 0, r = 5, b = 0),
      panel.grid.minor = element_blank()) 


pdf(paste0(outdatadir, "Volcano.pdf"), width=2.5, height=3.5, family="ArialMT")
  p 
dev.off()  
```


```{r}
upreguliuoti <- d[logFC > log2(FC) & FDR <= pval, ] %>% nrow()
downreguliuoti <- d[logFC < (-1)*log2(FC) & FDR <= pval, ] %>% nrow()
set.seed(1234)
dd <- d %>% 
  as.data.table() %>%
  .[, .(rn, logFC, FDR)] %>%
  .[, forma := "p. adj > 0"]  %>% 
  .[FDR == 0, forma := "p. adj == 0"]  %>% 
  .[FDR == 0, FDR := 1e-297]  %>% 
  .[logFC > 6, forma := "p. adj == 0"] %>% 
  .[logFC > 6, logFC := 5.9] %>% 
  .[, yval := (-1) * log10(FDR)] %>%
  .[, type := "Non. sig."] %>%
  .[abs(logFC) >= log2(FC), type := "Non. sig."] %>%
  .[FDR <= pval, type := "Non. sig."] %>%
  .[(abs(logFC) >= log2(FC)) & (FDR <= pval), type := "FC & p-adj."] %>%
  .[, type := factor(type, levels=c("Non. sig.", "FC & p-adj."))] %>% 
  merge(., tikslines[, .(LocusTag, fullGene)], by.x="rn", by.y="LocusTag", all.x=TRUE)  %>% 
  .[, fullGene := gsub(".*,", "", fullGene)] 

tikslines <- tikslines  %>% 
  .[, fullGene := gsub(".*,", "", fullGene)] 


nudgeY <- rep(0, nrow(dd))
nudgeY[dd$fullGene == tikslines$fullGene[1]] <- -0.2   #llmg_0439
nudgeY[dd$fullGene == tikslines$fullGene[2]] <- 0.1   # ps350
nudgeY[dd$fullGene == tikslines$fullGene[3]] <- 1   # ps356
nudgeY[dd$fullGene == tikslines$fullGene[4]] <- 0.1  # cluA 
nudgeY[dd$fullGene == tikslines$fullGene[7]] <- -0.65   # arcD1
nudgeY[dd$fullGene == tikslines$fullGene[9]] <- 1   # arcA   
nudgeY[dd$fullGene == tikslines$fullGene[10]] <- 1  # rcfB 
nudgeY[dd$fullGene == tikslines$fullGene[11]] <- -0.5  # llmg 2513

nudgeY[dd$fullGene == tikslines$fullGene[8]] <- -0.1   # argF/B
nudgeY[dd$fullGene == tikslines$fullGene[5]] <- 1   # arrC, arcC2 
nudgeY[dd$fullGene == tikslines$fullGene[6]] <- 0   # arrC, arcC1

nudgeX <- rep(0, nrow(dd))
nudgeX[dd$fullGene == tikslines$fullGene[1]] <-  0.2   #llmg_0439
nudgeX[dd$fullGene == tikslines$fullGene[2]] <-  -1.2   # ps350
nudgeX[dd$fullGene == tikslines$fullGene[3]] <- -1   # ps356
nudgeX[dd$fullGene == tikslines$fullGene[4]] <- 1   # cluA 
nudgeX[dd$fullGene == tikslines$fullGene[7]] <- 1   # arcD1
nudgeX[dd$fullGene == tikslines$fullGene[9]] <- -0.5   # arcA   
nudgeX[dd$fullGene == tikslines$fullGene[10]] <- 0  # rcfB 
nudgeX[dd$fullGene == tikslines$fullGene[11]] <- -0.9  # llmg 2513
nudgeX[dd$fullGene == tikslines$fullGene[8]] <- -1.5   # argF/B
nudgeX[dd$fullGene == tikslines$fullGene[5]] <- 1.5   # arrC, arcC2 
nudgeX[dd$fullGene == tikslines$fullGene[6]] <- 1.5   # arrC, arcC1


p_mod <- ggplot(dd, aes(logFC, yval, colour=type, label=fullGene, shape=forma)) +
    geom_point(size=0.55) +
    theme_bw() +
    geom_text_repel(
                    box.padding = -0.05,  
                    min.segment.length = 0, 
                    colour = "black", 
                    size = 2,
                    direction = "both",
                    segment.size = 0.25, 
                    max.overlaps = Inf,
                    vjust = 1.5,      
                    nudge_y = nudgeY, 
                    nudge_x = nudgeX
                    ) +
    ylab("-log10(FDR)") +
    xlab("log2FoldChange") + 
    scale_colour_manual(values=c(brewer.pal(n=3,"Dark2")[8], brewer.pal(n=8,"Dark2")[1])) +
    guides(color = guide_legend(ncol = 2)) +
    coord_cartesian(xlim=c(-5.5, 5.5)) +
    scale_shape_manual(values=c(17, 16)) +
    guides(color = guide_legend(nrow = 2), shape = guide_legend(nrow = 2)) +
    annotate("text", x = -3.9, y = 0.75, 
             label = paste0(downreguliuoti, " down", "\u002D", "\nregulated"), 
             lineheight = 0.9,
             color = brewer.pal(n=8,"Dark2")[1], size = 2.75) + 
    annotate("text", x = 3.9, y = 0.75, 
             label = paste0(upreguliuoti, " up", "\u002D", "\nregulated"), 
             lineheight = 0.9,
             color = brewer.pal(n=8,"Dark2")[1], size = 2.75) +           
    theme(legend.position ="none", 
      axis.text = element_text(size = 8, colour="black"),  
      axis.title = element_text(size = 8, colour="black"),  
      plot.margin = margin(t = 7, l = 0, r = 5, b = 0),
      panel.grid.minor = element_blank()) 
p_mod      
```

```{r}
d <- readRDS(paste0("output/code/DE_analysis_edgeR/", "/KEGG_GSEA.RDS"))
d@result$Description <- gsub(" - Lactococcus lactis subsp. cremoris MG1363", "", d$Description)
p1 <- clusterProfiler::dotplot(d, color="qvalue") +
  theme_bw() +
    #guides(color = guide_legend(ncol = 2)) +
    coord_cartesian(xlim=c(0.0001, 1)) +
    scale_size(range = c(1, 5), 
              limits=c(5, 50),
              breaks=c(10, 20, 30, 40, 50),
              ) +
    scale_color_continuous(low="red", high="blue", 
            name = "q-value",
            breaks = c(0.00, 0.025, 0.05),
            guide=guide_colorbar(reverse=TRUE), limits=c(0, 0.055)) +
    theme(legend.position ="right", 
      axis.text = element_text(size = 8, colour="black"),  
      axis.title = element_text(size = 8, colour="black"),  
      legend.text = element_text(size = 7, margin = margin(l = 0, r = 0, t = 0, b = 0)), 
      plot.margin = margin(t = 1, r = 0, b = 0, l=5),
      legend.margin = margin(t = 0, b = 0, r = 0, l=0),
      legend.spacing = unit(0.1, "lines"),
      legend.title = element_text(size=8, margin = margin(b = 0)),  # Reduce space below the legend title
      panel.grid.minor = element_blank(),
      plot.title = element_blank(),
      plot.subtitle = element_blank()) 
p1 

pdf(paste0(outdatadir, "GSEA_enrichment.pdf"), width=4, height=3)
  p1 
dev.off()  
```

```{r}
px <- p | p1

pdf(paste0(outdatadir, "sLLM1042_VolcanoGSEA.pdf"), width=7.01, height=2.97, family="ArialMT")
  px
dev.off() 
```

```{r}
px <- p_mod | p1

pdf(paste0(outdatadir, "sLLM1042_VolcanoGSEA_mod.pdf"), width=7.01, height=2.97, family="ArialMT")
  px
dev.off() 
```